tests/test_0652-tests-of-complex-numbers.py

```diff
@@ -0,0 +1,214 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+from __future__ import absolute_import
+
+import json
+
+import pytest  # noqa: F401
+import numpy as np  # noqa: F401
+import awkward as ak  # noqa: F401
+
+
+def test_from_iter():
+    assert ak.from_iter([1 + 1j, 2 + 2j, 3 + 3j]).tolist() == [1 + 1j, 2 + 2j, 3 + 3j]
+    assert ak.from_iter([[1 + 1j, 2 + 2j], [], [3 + 3j]]).tolist() == [
+        [1 + 1j, 2 + 2j],
+        [],
+        [3 + 3j],
+    ]
+
+    # First encounter of a complex number should promote previous integers and
+    # reals into complex numbers:
+    assert ak.from_iter([1, 2.2, 3 + 3j]).tolist() == [1.0 + 0j, 2.2 + 0j, 3.0 + 3j]
+    assert ak.from_iter([1, 3 + 3j]).tolist() == [1.0 + 0j, 3.0 + 3j]
+
+    # Just as the first encounter of a real number promotes previous integers
+    # into reals:
+    assert str(ak.from_iter([1, 2.2]).type) == "2 * float64"
+    assert ak.from_iter([1, 2.2]).tolist() == [1.0, 2.2]
+    builder = ak.ArrayBuilder()
+    str(ak.type(builder)) == "0 * unknown"
+    builder.integer(1)
+    str(ak.type(builder)) == "1 * int64"
+    builder.real(2.2)
+    str(ak.type(builder)) == "2 * float64"
+
+    # For that matter, ArrayBuilder is missing a high-level interface to complex:
+    builder.complex(3 + 3j)
+    str(ak.type(builder)) == "3 * complex128"
+
+
+def test_from_json():
+    array = ak.from_json('[{"r": 1.1, "i": 1.0}, {"r": 2.2, "i": 2.0}]')
+    assert array.tolist() == [
+        {"r": 1.1, "i": 1.0},
+        {"r": 2.2, "i": 2.0},
+    ]
+    array = ak.from_json(
+        '[{"r": 1.1, "i": 1.0}, {"r": 2.2, "i": 2.0}]', complex_record_fields=("r", "i")
+    )
+    assert (array["r"] + array["i"]*1j).tolist() == [(1.1 + 1j), (2.2 + 2j)]
+
+    # Somewhere in from_json, a handler that turns integer record fields into
+    # parts of a complex number is missing.
+    array = ak.from_json(
+        '[{"r": 1, "i": 1}, {"r": 2, "i": 2}]', complex_record_fields=("r", "i")
+    )
+    assert (array["r"] + array["i"]*1j).tolist() == [(1 + 1j), (2 + 2j)]
+
+    # # This should fail with some message like "complex number fields must be numbers,"
+    # # not "called 'end_record' without 'begin_record' at the same level before it."
+    # with pytest.raises(ValueError) as err:
+    #     array = ak.from_json(
+    #         '[{"r": [], "i": 1}, {"r": [1, 2], "i": 2}]',
+    #         complex_record_fields=("r", "i"),
+    #     )
+    #     assert array["r"].type == array["i"].type
+    # assert "Complex number fields must be numbers" in str(err)
```

Comment:
This test can probably be turned on again, now that the error message is handled in Python.

And also, it closes what I called above an "open question" about how to handle fields that aren't one-dimensional NumpyArrays. When I wrote that comment, I thought it should be an error. Sure, it could be an error. I'll agree with myself.

Reply:
done